### PR TITLE
GSoC - Curve Fitting

### PR DESCRIPTION
**Pull request by Pranavan135**
Do not merge. Tentative PR to get feedbacks for my work.

### CODE REVIEW

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/rule/RuleTasksDelegator.java [hunk 1]
```diff
@@ -357,6 +378,12 @@ public double getLoadAveragePredictedValue(ClusterInstanceContext clusterInstanc
         }
     }
 
+    public double getLoadAveragePredictedValue(){
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/rule/RuleTasksDelegator.java]: Minor fix. Format the code properly with spaces.

File: components/org.apache.stratos.messaging/src/main/java/org/apache/stratos/messaging/event/health/stat/CurveFinderOfLoadAverageEvent.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.stratos.messaging.event.health.stat;
+
+import org.apache.stratos.messaging.event.Event;
+
+
+public class CurveFinderOfLoadAverageEvent extends Event{
```
Comment (lahirus) [inline components/org.apache.stratos.messaging/src/main/java/org/apache/stratos/messaging/event/health/stat/CurveFinderOfLoadAverageEvent.java]: Minor formatting issue mentioned earlier.

File: extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/CurveFinderWindowProcessor.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+import java.util.concurrent.TimeUnit;
+
+@SiddhiExtension(namespace = "stratos", function = "curveFitting")
+public class CurveFinderWindowProcessor extends WindowProcessor implements RunnableWindowProcessor{
+
+    static final Logger log = Logger.getLogger(CurveFinderWindowProcessor.class);
+    /**
+     * alpha is used as the smoothing constant in exponential moving average
+     */
+    public static final double ALPHA = 0.8000;
```
Comment (R-Rajkumar) [inline extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/CurveFinderWindowProcessor.java]: make this ALPHA configurable.

File: extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/CurveFitter.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+         */
+        final PolynomialCurveFitter polynomialCurveFitter = PolynomialCurveFitter.create(2);
+        final double[] coefficients = polynomialCurveFitter.fit(weightedObservedPoints.toList());
+
+        log.info("Coefficients a : " + coefficients[0] + " b : " + coefficients[1] + " c : " + coefficients[2]);
+
+        return coefficients;
+    }
+
+}
```
Comment (R-Rajkumar) [inline extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/CurveFitter.java]: add the licence header to this class

File: extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/ExponentialMovingAverage.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        double newValue = CurveFinderWindowProcessor.ALPHA * value + (1.0 - CurveFinderWindowProcessor.ALPHA)*value;
+        this.value = newValue;
+
+        return newValue;
+    }
+
+    public Double getValue() {
+        return value;
+    }
+}
```
Comment (R-Rajkumar) [inline extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/ExponentialMovingAverage.java]: add the licence header to this class

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/rule/RuleTasksDelegator.java (hunk 1 above)
Comment (Pranavan135) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/rule/RuleTasksDelegator.java]: Resolved

File: components/org.apache.stratos.messaging/src/main/java/org/apache/stratos/messaging/event/health/stat/CurveFinderOfLoadAverageEvent.java (hunk 2 above)
Comment (Pranavan135) [inline components/org.apache.stratos.messaging/src/main/java/org/apache/stratos/messaging/event/health/stat/CurveFinderOfLoadAverageEvent.java]: Resolved

File: extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/CurveFitter.java (hunk 4 above)
Comment (Pranavan135) [inline extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/CurveFitter.java]: License Header Added

File: extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/ExponentialMovingAverage.java (hunk 5 above)
Comment (Pranavan135) [inline extensions/cep/stratos-cep-extension/src/main/java/org/apache/stratos/cep/extension/ExponentialMovingAverage.java]: License Header Added

File: components/org.apache.stratos.messaging/src/main/java/org/apache/stratos/messaging/event/health/stat/MemberCurveFinderOfLoadAverageEvent.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package org.apache.stratos.messaging.event.health.stat;
+
+import org.apache.stratos.messaging.event.Event;
+
+/**
```
Comment (lahirus) [inline components/org.apache.stratos.messaging/src/main/java/org/apache/stratos/messaging/event/health/stat/MemberCurveFinderOfLoadAverageEvent.java]: Remove author tag

File: components/org.apache.stratos.messaging/src/main/java/org/apache/stratos/messaging/message/processor/health/stat/CurveFinderOfLoadAverageMessageProcessor.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("%s event processor notified listeners ... ", type));
+            }
+            return true;
+        } else {
+            if (nextProcessor != null) {
+                return nextProcessor.process(type, message, object);
+            } else {
+                throw new RuntimeException(String.format("Failed to process health stat message using available message processors: [type] %s [body] %s", type, message));
```
Comment (lahirus) [inline components/org.apache.stratos.messaging/src/main/java/org/apache/stratos/messaging/message/processor/health/stat/CurveFinderOfLoadAverageMessageProcessor.java]: Have custom exceptions, not the RuntimeException

**Comment by Pranavan135** [comment]
New PR created with updated changes. Closing this one.

**Pull request closed without merging**